[Requirements] Add a lock-file for mlrun-api Python dependencies

jond01 opened the pull request:
Use uv 0.5.5: https://docs.astral.sh/uv/.

Implements [ML-8383](https://iguazio.atlassian.net/browse/ML-8383).

Apart from locking the dependencies for reproducibility, it also:
* Significantly accelerates the `make api` **build time** to 30% of the time with pip (no cache).
* Reduces the **image size** to 68% the size with pip.

The `mlrun/mlrun-api` Docker image figures:

| tool                 | pip  | uv   |
|----------------------|------|------|
| build time (minutes) | 4:55 | 1:29 |
| size (GB)            | 4.37 | 2.96 |

From my local testing - the Docker image works perfectly fine.

[ML-4648]: https://iguazio.atlassian.net/browse/ML-4648?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[ML-8383]: https://iguazio.atlassian.net/browse/ML-8383?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

jond01 committed:
Add a make target to create/update the mlrun-api lock file

jond01 committed:
Add the lock file

```sh
make update-api-deps
```

jond01 committed:
Set the Python version and choose a universal resolution

jond01 committed:
Update the lock file

jond01 committed:
Add `.venv` to `.dockerignore`

See:
https://docs.astral.sh/uv/guides/integration/docker/#installing-a-project

jond01 committed:
Update mlrun-api/Dockerfile to use uv

jond01 committed:
Fix `test_requirements.py` - ignore lock files

jond01 committed:
Add hashes to the lock file

jond01 committed:
Move some of the config to pyproject.toml

Update the lockfile

jond01 committed:
Require hashes in the Dockerfile

liranbg commented on Makefile:
```diff
@@ -775,3 +775,15 @@ ifdef MLRUN_DOCKER_CACHE_FROM_TAG
+	dockerfiles/mlrun-api/requirements.txt \
+	--python-version $(MLRUN_PYTHON_VERSION) \
+	--universal \
+	--generate-hashes \
+	-U -o dockerfiles/mlrun-api/locked-requirements.txt
```
add condition to update lock file when adding specific requirement

jond01 replied:
Done in 218007b

jond01 committed:
Verify the running uv version in the make target

jond01 committed:
Update the make target flags

jond01 committed:
Require hashes only for the lock file

jond01 committed:
Use `cut` instead of `jq` to extract uv's version

jond01 committed:
Update the lock file without `--upgrade`

jond01 committed:
Add `MLRUN_UV_UPGRADE_FLAG`

jond01 committed:
Merge branch 'development' into task/lockfile-for-mlrun-api-python-dependencies

jond01 committed:
Upgrade mlrun-api deps

jond01 committed:
Merge branch 'development' into task/lockfile-for-mlrun-api-python-dependencies

jond01 committed:
Upgrade uv to version 0.5.3

jond01 committed:
Remove the verify-hashes setting

It's true by default in 0.5.3:
https://github.com/astral-sh/uv/releases/tag/0.5.3

jond01 committed:
Update the lock file

jond01 committed:
Merge remote-tracking branch 'upstream/development' into task/lockfile-for-mlrun-api-python-dependencies

# Conflicts:
#	dockerfiles/mlrun-api/Dockerfile

jond01 committed:
Update the lockfile

jond01 committed:
Merge branch 'development' into task/lockfile-for-mlrun-api-python-dependencies

# Conflicts:
#	dockerfiles/mlrun-api/Dockerfile

jond01 committed:
Update lock file

jond01 committed:
Add a section to the contributing file

jond01 committed:
Merge branch 'development' into task/lockfile-for-mlrun-api-python-dependencies

jond01 committed:
Update lock

jond01 committed:
Upgrade uv to 0.5.5

jond01 committed:
Merge branch 'development' into task/lockfile-for-mlrun-api-python-dependencies

jond01 committed:
Fix the contributing doc

* Add "uv" to ignored (allowed) words
* Remove duplicate "the"

quaark commented on Makefile:
```diff
@@ -782,3 +786,23 @@ ifdef MLRUN_DOCKER_CACHE_FROM_TAG
+	fi; \
+	}
+
+.PHONY: update-mlrun-api-deps
+update-api-deps: verify-uv-version ## Update mlrun-api locked requirements file
```
The .PHONY name doesn't match the rule name.
Anyway, maybe call it more explicitly `update-mlrun-api-deps-lock` (just to have 'lock' in the name to make it more clear what this does)

jond01 replied:
Renamed to `upgrade-mlrun-api-deps-lock` and fixed the .PHONY name in 4e98e41

quaark reviewed (changes requested):
Looks great!
Minor naming issue/suggestion in the makefile

liranbg commented on Makefile:
```diff
@@ -782,3 +786,23 @@ ifdef MLRUN_DOCKER_CACHE_FROM_TAG
+verify-uv-version:
+	@{ \
+	uv_version=$$(uv version | cut -d' ' -f2); \
+	result=$$(pysemver compare $$uv_version $(MLRUN_UV_VERSION)); \
```
```suggestion
	result=$$(python -m semver compare $$uv_version $(MLRUN_UV_VERSION)); \
```

While this is a suggestion, the reasons for it
1. explicitness
2. indicates exactly which python runs the cli command

jond01 replied:
Done in b5db017

jond01 committed:
Merge branch 'development' into task/lockfile-for-mlrun-api-python-dependencies

jond01 committed:
Change command to `python -m semver` instead of `pysemver`

jond01 committed:
Rename phony target to `upgrade-mlrun-api-deps-lock`

jond01 committed:
Upgrade deps
